kep/4986-sakkara-cluster-topology/README.md

```diff
@@ -0,0 +1,536 @@
+- [pod topology spread constraints](https://kubernetes.io/docs/concepts/scheduling-eviction/topology-spread-constraints/): a very narrow form of topology constraint of a single pod
+
+Lacking is a scheduler that is both topology-aware at the cluster level, with a set of rich topological constraints, and places/preempts whole groups.
+
+### Goals
+
+Develop a scheduler plugin which places a group of pods in a cluster with a given tree topology, satisfying group placement constraints at all levels of the hierarchy.
+
+### Non-Goals
+
+- Discovery of cluster topology
+- Migration of pods
+- Elastic groups with changing size
+
+## Proposal
+
+### User Stories
+
+1- **User A:**
+
+Want to deploy a job with 8 pods, equally partitioned into two partitions, where each partition is placed in separate rack, and the pods belonging to a partition are packed on nodes within the rack. Partitioning improves availability, and packing improves performance.
+
+![before-1](images/use-cases/before-1.png)
+
+![after-1](images/use-cases/after-1.png)
+
+2- **User B:**
+
+Want to place a job with 6 pods packed on same physical server, as much as possible, where pods on the same server are spread across nodes in that server, but in the range [2,4]. Having all pods on the same physical server improves communication among the pods. And, having less than 2 pods in a node spreads the pods too thinly, and having more than 4 may leads to congestion.
+
+![before-2](images/use-cases/before-2.png)
+
+![after-2](images/use-cases/after-2.png)
+
+### Examples
+
+![topology](images/example/topology.png)
+
+![group](images/example/group.png)
+
+![placement](images/example/placement.png)
+
+![result-group](images/example/result-group.png)
+
+![result-pod](images/example/result-pod.png)
+
+![premption before](images/preemption/before.png)
+
+![premption after-1](images/preemption/after-1.png)
+
+![premption after-2](images/preemption/after-2.png)
+
+### Notes/Constraints/Caveats
+
+### Risks and Mitigations
+
+## Design Details
+
+![architecture](images/architecture.png)
+
+![plugin state transitions](images/state-transitions.png)
+
+### Cluster topology
+
+The cluster topology is specified as a tree where the leaves are the nodes in the cluster.
+There may be one or more levels in the tree, where nodes are at level 0.
+For example, level 1 may represent servers, level 2 racks, level 3 rows, and so on.
+If no cluster topology is specified, Sakkara assumes a flat topology with only one level, i.e. a tree with a root and all nodes in the cluster as leaves, with two resources: CPU and memory.
+
+The tree is specified in a ConfigMap with a label `"sakkara.topology":`.
+It may be updated at any time. Updates take effect during the following group placement cycle.
+
+An example with a two-level tree of racks and nodes follows.
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
```

Comment:
Not a blocker for the KEP, but I highly recommend to start with CRD since day 1.

Reply:
Agreed. Should we introduce a new CR or extend the PodGroup?